### PR TITLE
[NeMo-UX] Add distributed checkpointing unit tests

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/lightning/test_dist_ckpt.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from nemo.collections import llm
+from nemo.lightning.io.pl import MegatronCheckpointIO
+from nemo.utils.callbacks.dist_ckpt_io import AsyncFinalizableCheckpointIO, AsyncFinalizerCallback
+
+
+def _get_nlp_strategy_without_optimizer_state():
+    strategy = nl.MegatronStrategy(
+        enable_nemo_ckpt_io=False,
+    )
+    # this ensures optimizer sharded state creation is skipped
```
Comment (mikolajblaz) [inline tests/lightning/test_dist_ckpt.py:20]: Did you test if with the new strategy testing the optimizer save works?

Previously it errored because of some parallel state misconfiguration I believe

File: tests/lightning/test_dist_ckpt.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            max_steps=2,
+            default_root_dir=str(tmp_path),
+        )
+
+        trainer.fit(model, data)
+
+        assert isinstance(trainer.strategy.checkpoint_io, MegatronCheckpointIO)
+        # Ckpt path doesn't contain the .ckpt suffix
+        ## TODO: make more generic
+        ## why does this path include "lightning logs" while original test does not?
```
Comment (mikolajblaz) [inline tests/lightning/test_dist_ckpt.py:68]: This looks like some generic PTL path, why `default_root_dir=str(tmp_path)` is not picked up?

File: tests/lightning/test_dist_ckpt.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        ## reset the model and data and train with sync checkpointing
+        model, data = get_model_and_data()
+        async_test_trainer = pl.Trainer(
+            devices=1,
+            enable_checkpointing=True,
+            logger=False,
+            max_steps=2,
+            strategy=_get_nlp_strategy_without_optimizer_state(),
+            plugins=[async_checkpoint_io],  ## error is not specific to async checkpointing
```
Comment (mikolajblaz) [inline tests/lightning/test_dist_ckpt.py:117]: > error is not specific to async checkpointing
what's this comment about?

File: tests/lightning/test_dist_ckpt.py (hunk 2 above)
Comment (ashors1) [reply tests/lightning/test_dist_ckpt.py]: It was because I hadn't disabled logging. I just fixed the issue.

File: tests/lightning/test_dist_ckpt.py (hunk 1 above)
Comment (ashors1) [reply tests/lightning/test_dist_ckpt.py]: The tests still pass if I allow the optimizer state to be created. Is that sufficient?

Comment (mikolajblaz) [reply tests/lightning/test_dist_ckpt.py]: yes thanks

**Review by mikolajblaz — APPROVED** [review]
